### PR TITLE
Add logo option to Endroid's provider

### PR DESCRIPTION
**Pull request by igorsantos07**
I know you suggested more constructor arguments, but after implementing and trying to use, I saw it wouldn't make much sense... Also, I understood you wanted to keep the original constructor as-is, with the logo being a deliberately and rare optional choice.

Thus, the method suits just fine, keeping everything at the same place, being less verbose and less bloated (otherwise a common call + logo would need to skip-guess all the default parameters, such as colors, margin, correction level...).

> P.S.: I'd suggest creating a new release! The two external providers are only available in `master`.

### CODE REVIEW

**Comment by RobThree** [comment]
The crux of [my suggestion](https://github.com/RobThree/TwoFactorAuth/issues/56#issuecomment-765727021), though, was to inherit/extend instead of change. I'd really prefer that actually.

**Comment by igorsantos07** [comment]
Will do, then. But why do you think yet another class is better, just to tweak a property of the existing QR Code generator? Not to mention the verbosity of the class name...

**Comment by RobThree** [comment]
> But why do you think yet another class is better

Because most people won't need that functionality

> just to tweak a property of the existing QR Code generator?

Because it makes using the class more complex than needed (though, granted, it's not rocketscience)

> Not to mention the verbosity of the class name...

As I said, I'm open to better names ;-)

**Comment by igorsantos07** [comment]
There's definitely no better name than `EndroidQrCodeWithLogoProvider`, since it follows the current pattern.

But most people won't change the background color from white (or shouldn't, as that affects readability), yet it's the first parameter on the constructor. That's the reason I added it as a method - who needs it, call it, who doesn't, just ignores it. Or are you concerned about an extra method in the class's memory?

But there's one practical side-effect I'm not sure how to deal with, without making the code more complex - check code comment.

File: lib/Providers/Qr/EndroidQrCodeProvider.php
Comment (igorsantos07) [inline lib/Providers/Qr/EndroidQrCodeProvider.php:55]: How would you suggest to add this piece of code in an extended class? Create _another_ protected method to instantiate the basic object, then use it in `getQRCodeImage()`?

**Comment by RobThree** [comment]
> There's definitely no better name than EndroidQrCodeWithLogoProvider, since it follows the current pattern.

Maybe. Not sure.

> Or are you concerned about an extra method in the class's memory?

No, I'm weighing [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and possible confusion for a method or constructor arguments that does something pretty uncommon.

> How would you suggest to add this piece of code in an extended class? Create another protected method to instantiate the basic object, then use it in getQRCodeImage()?

I'm not sure I understand the question but that may be because it's 4:20AM over here currently. I'll look into it this week and get back to you.

**Comment by igorsantos07** [comment]
Gotcha. I guess we can agree to disagree in this situation :)

And definitely, that's not a good time to review code :joy: I'll submit my best suggestion and you can check later. Good night!

**Comment by RobThree** [comment]
> Gotcha. I guess we can agree to disagree in this situation :)

Don't get me wrong, your effort and feedback is appreciated, and you do make a point!

> Good night!

Thanks!

**Comment by igorsantos07** [comment]
Yep! I was wondering if you were being stubborn, but in the end, it's just different ways to program :) That's why I opened the issue initially. Not to mention I never had a PR replied THAT FAST. I'm used to being stuck with custom branches on my composer files :sweat:

File: README.md
Comment (igorsantos07) [inline README.md:132]: [-[`EndroidQrCodeProvider`](lib/Providers/Qr/EndroidQrCodeProvider.php)-]{+[`EndroidQrCodeProvider`](lib/Providers/Qr/EndroidQrCodeProvider.php), [`EndroidQrCodeWithLogoProvider`](lib/Providers/Qr/EndroidQrCodeWithLogoProvider.php)+}

**Comment by RobThree** [comment]
> I was wondering if you were being stubborn

I'm not denying that 😇

I'm off to bed now. Will get back to you! Again, thanks! 👊

File: lib/Providers/Qr/EndroidQrCodeProvider.php
Comment (igorsantos07) [inline lib/Providers/Qr/EndroidQrCodeProvider.php:42]: So child classes can tweak the object without messing with how the QRCode image is returned.

File: lib/TwoFactorAuth.php
Comment (igorsantos07) [inline lib/TwoFactorAuth.php:85]: Stumbled upon this when I type-hinted in my internal method with `int`... I discovered the hard way that `hash_equals('123', 123)` is false.

**Comment by igorsantos07** [comment]
I ended up including some extra stuff in the same PR. Let me know if you prefer a second PR with the unrelated details :)

File: README.md
Comment (igorsantos07) [inline README.md:100]: Since I was going to mention an explicit requirement for the $code argument, I moved it all to a separate paragraph, coming before the return (you first fill in the args, and then deal with the return, right?)

File: README.md
Comment (igorsantos07) [inline README.md:102]: And then, the following paragraph starts where the next two do: with the extra arguments' description.

File: lib/TwoFactorAuth.php
Comment (igorsantos07) [inline lib/TwoFactorAuth.php:87]: I also added the `string $code` to a docblock to make it clear it won't work if casted - besides losing the zeroes, `hash_equals('123', 123)` returns `false`

**Comment by VictorVolpe** [comment]
Great PR @igorsantos07
I will use it right now!

**Comment by igorsantos07** [comment]
@RobThree / @willpower232 I was wondering if this got stuck because it was just forgotten or because of the discussions whether extra providers will be separated into their own packages. Or maybe if I need to do something else (other than maybe fix the conflicts, but I guess maintainers would be better at this?)

**Comment by willpower232** [comment]
it would indeed be good to get a steer from @RobThree about whether the bacon and endroid providers should be evicted to their own repos and where those should be. IAlso  need to push a lot of docblocks and a couple of fixes I've spotted.

**Comment by RobThree** [comment]
I think moving to separate libraries would be a 2.0 move, unless someone disagrees?

**Comment by willpower232** [comment]
makes sense but what should happen to the ones already merged :sweat_smile:

**Comment by igorsantos07** [comment]
I understand we wouldn't be including extra providers until v2, at which point all providers would have their own, separate repos.

However, I don't think this "extraction" move should hinder improvements on the current providers. Otherwise, they'll just get stalled "waiting for the perfect moment". That sort of how I killed an open-source project I adopted a couple of years ago - I was holding on extra external changes and fixes while I was preparing to write tests for it... The tests never came, and the package just felt into disrepair.

**Comment by RobThree** [comment]
> I understand we wouldn't be including extra providers until v2, at which point all providers would have their own, separate repos.
> 
> However, I don't think this "extraction" move should hinder improvements on the current providers. Otherwise, they'll just get stalled "waiting for the perfect moment". That sort of how I killed an open-source project I adopted a couple of years ago - I was holding on extra external changes and fixes while I was preparing to write tests for it... The tests never came, and the package just felt into disrepair.

You are absolutely right. However, I'm currently very busy and I really need the little rest I get. Maybe @willpower232 will be able to help out or, with his OK, I can add you as contributor too.

**Comment by willpower232** [comment]
I like the way this has been done and would be happy to merge, if @RobThree is happy for endroid and bacon to be part of a release that is. I guess they could stand for examples of providers and if people want to write more in the future, you want to recommend they run their own package?

@igorsantos07 looks like there is a conflict in TwoFactorAuthTest probably I've been working on it

**Review by RobThree — APPROVED** [review]

**Comment by RobThree** [comment]
> if @RobThree is happy for endroid and bacon to be part of a release that is

Sure, why not 😉Go ahead. I think it's good 👍

**Comment by willpower232** [comment]
all done! :tada:

File: lib/TwoFactorAuth.php
Comment (igorsantos07) [reply lib/TwoFactorAuth.php:83]: Sorry, non native here 😂

**Pull request merged**
